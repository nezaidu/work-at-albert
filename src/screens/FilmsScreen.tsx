import React from 'react';
import {
  NavigationComponentProps,
  NavigationFunctionComponent,
} from 'react-native-navigation';
import {
  Film,
  SortingDirection,
  useFilmsQuery,
  useMetaQuery,
  MetaDocument,
} from '../generated/type';
import Films from '../components/Films';
import {cache} from '../core';
import moment from 'moment';
import { fonts } from '../../ui';

type UseFilms = (d: SortingDirection) => Film[];

export const useFilms: UseFilms = (direction: SortingDirection) => {
  const {data} = useFilmsQuery({
    fetchPolicy: 'cache-and-network'
  });

  let films = (data?.allFilms?.films || []) as Film[];
  const newFilms = [...films];

  switch (direction) {
    case SortingDirection.Asc:
      return newFilms.sort(
        (f1, f2) =>
          moment(f1.releaseDate).unix() - moment(f2.releaseDate).unix(),
      );
    case SortingDirection.Desc:
      return newFilms.sort(
        (f1, f2) =>
          moment(f2.releaseDate).unix() - moment(f1.releaseDate).unix(),
      );
    default:
      return newFilms;
  }
};

type UseSorting = () => {
  sortingDirection: SortingDirection;
  toggleSorting: () => SortingDirection;
};

export const useSorting: UseSorting = () => {
  const { data } = useMetaQuery({
    fetchPolicy: 'cache-only',
  });

  const sortingDirection = data?.meta?.filmsSortingDirection || SortingDirection.Asc;

  const toggleSorting = () => {
    const newDirection =
      sortingDirection === SortingDirection.Asc
        ? SortingDirection.Desc
        : SortingDirection.Asc;

    cache.writeQuery({
      query: MetaDocument,
      data: {
        meta: {
          filmsSortingDirection: newDirection,
        }
      }
    });
    return newDirection;
  };

  return {
    sortingDirection,
    toggleSorting,
  };
};

const FilmsScreen: NavigationFunctionComponent<NavigationComponentProps> = () => {
  const {sortingDirection, toggleSorting} = useSorting();

  const films = useFilms(sortingDirection);

  return (
    <Films
        data={films}
        toggleSorting={toggleSorting}
        sortingDirection={sortingDirection}
    />
  )
};


FilmsScreen.options = {
  topBar: {
    background: {
      color: '#000000',
    },
    title: {
      text: "Star Wars",
      fontFamily: fonts.distant,
      fontSize: 22,
      fontWeight: '900',
      color: '#F5ED17',
    }
  }
}

export default FilmsScreen;
