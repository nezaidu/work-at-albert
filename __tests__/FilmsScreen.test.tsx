/**
 * @format
 */
/* global __DEV__ */
import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import {render} from '@testing-library/react-native';
import { renderHook, cleanup, act } from '@testing-library/react-hooks';
import FilmsScreen, { useFilms, useSorting } from '../src/screens/FilmsScreen';
import initClient from '../src/core';
import filmsJSON from '../__mocks__/films.json';
import { Film, MetaDocument, SortingDirection } from '../src/generated/type';

initClient();

const films: [Film] = filmsJSON as unknown as [Film];

const mockedType = require('../src/generated/type');

mockedType.useFilmsQuery = () => {
  return films;
};

mockedType.useMetaQuery = () => {
  return {
    data: {
      meta: {
        filmsSortingDirection: SortingDirection.Asc,
      },
    },
  };
};

it('fetches films', () => {
  const {toJSON } = render(<FilmsScreen />);

  const tree = toJSON();

  expect(tree).toMatchSnapshot();
});

it('sorts films by release date ASC', () => {
  const sortedFilms = useFilms(SortingDirection.Asc);
  expect(sortedFilms[0].releaseDate).toMatch("1977-05-25");
});

it('sorts films by release date DESC', () => {
  const sortedFilms = useFilms(SortingDirection.Desc);
  expect(sortedFilms[0].releaseDate).toMatch("2005-05-19");
});

it('toggles sorting direction', () => {
  const { sortingDirection, toggleSorting } = useSorting();

  let newSorting;
  act(() => {
    newSorting = toggleSorting();
  });

  expect(sortingDirection).not.toMatch(newSorting);
});
