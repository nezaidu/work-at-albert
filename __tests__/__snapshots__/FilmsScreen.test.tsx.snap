// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches films 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "episodeID": 4,
        "id": "ZmlsbXM6MQ==",
        "releaseDate": "1977-05-25",
        "title": "A New Hope",
      },
      Object {
        "episodeID": 5,
        "id": "ZmlsbXM6Mg==",
        "releaseDate": "1980-05-17",
        "title": "The Empire Strikes Back",
      },
      Object {
        "episodeID": 6,
        "id": "ZmlsbXM6Mw==",
        "releaseDate": "1983-05-25",
        "title": "Return of the Jedi",
      },
      Object {
        "episodeID": 1,
        "id": "ZmlsbXM6NA==",
        "releaseDate": "1999-05-19",
        "title": "The Phantom Menace",
      },
      Object {
        "episodeID": 2,
        "id": "ZmlsbXM6NQ==",
        "releaseDate": "2002-05-16",
        "title": "Attack of the Clones",
      },
      Object {
        "episodeID": 3,
        "id": "ZmlsbXM6Ng==",
        "releaseDate": "2005-05-19",
        "title": "Revenge of the Sith",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="toggle-sorting-button"
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 1,
                  "flexDirection": "row",
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "borderRadius": 1,
                    "marginLeft": -8,
                    "marginRight": -8,
                    "marginTop": -8,
                    "shadowColor": "#0339c9",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 1,
                    "shadowRadius": 12,
                  },
                ]
              }
            >
              <SvgMock />
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#817156",
                    "fontFamily": "Star Jedi Rounded",
                    "fontSize": 18,
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                    "lineHeight": 16,
                    "paddingBottom": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                    "paddingTop": 10,
                  },
                ]
              }
            >
              1977
              ...
              2005
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          4
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            A New Hope
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            1977
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          5
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            The Empire Strikes Back
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            1980
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          6
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            Return of the Jedi
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            1983
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          1
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            The Phantom Menace
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            1999
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          2
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            Attack of the Clones
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            2002
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#f5ed17",
            "borderColor": "#0339c9",
            "borderRadius": 1,
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 15,
            "opacity": 1,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 3,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#817156",
                "fontFamily": "Star Jedi Rounded",
                "fontSize": 22,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 18,
                "paddingBottom": 10,
                "paddingLeft": 10,
                "paddingRight": 10,
                "paddingTop": 10,
              },
            ]
          }
        >
          3
        </Text>
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "Star Jedi Rounded",
                  "fontSize": 20,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 18,
                  "paddingBottom": 0,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                },
              ]
            }
          >
            Revenge of the Sith
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#010103",
                  "fontFamily": "SerifGothic LT",
                  "fontSize": 16,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "lineHeight": 14,
                  "paddingLeft": 10,
                  "paddingTop": 3,
                },
              ]
            }
          >
            2005
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
